batcher/sweep_batch.go

```diff
@@ -0,0 +1,1330 @@
+			// A re-org has been detected. We set the batch's state
+			// back to open since our batch transaction is no longer
+			// present in any block. We can accept more sweeps and
+			// try to publish new transactions, at this point we
+			// need to monitor again for a new spend.
+			b.State = Open
+			b.log.Warnf("reorg detected")
+			return
+
+		case value := <-b.confRefreshChan:
+			// If the nonce we received from the channel is equal
+			// to this routine's nonce, this means that we are the
+			// first routine to arrive here. So we only exit if the
+			// nonce we read is different.
+			if value != nonce {
+				return
+			}
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// handleSpend handles a spend notification.
+func (b *Batch) handleSpend(ctx context.Context, spendTx *wire.MsgTx) error {
+	// As soon as a spend arrives we should no longer accept sweeps. Also,
+	// the closed state will cause the batch to start monitoring the
+	// confirmation of the batch transaction.
+	b.State = Closed
+
+	b.sweepsLock.Lock()
+
+	txHash := spendTx.TxHash()
+	b.BatchTxid = &txHash
+	b.BatchPkScript = spendTx.TxOut[0].PkScript
+	purgeList := make([]SweepRequest, 0, len(b.sweeps))
+	spendCounter := 0
+
+	// As a previous version of the batch transaction may get confirmed,
+	// which does not contain the latest sweeps, we need to detect the
+	// sweeps that did not make it to the confirmed transaction and feed
+	// them back to the batcher. This will ensure that the sweeps will enter
+	// a new batch instead of remaining dangling.
+	for _, sweep := range b.sweeps {
+		found := false
+
+		for _, txIn := range spendTx.TxIn {
+			if txIn.PreviousOutPoint == sweep.Outpoint {
+				found = true
+				spendCounter++
+			}
+		}
+
+		// If the sweep's outpoint was not found in the transaction's
+		// inputs this means it was left out. So we delete it from this
+		// batch and feed it back to the batcher.
+		if !found {
+			newSweep := sweep
+			delete(b.sweeps, sweep.SwapHash)
+			purgeList = append(purgeList, SweepRequest{
+				SwapHash: newSweep.SwapHash,
+				Outpoint: newSweep.Outpoint,
+				Value:    newSweep.Value,
+			})
+		}
+	}
+
+	b.sweepsLock.Unlock()
+
+	// If we have sweeps to purge, we send them back to the batcher.
+	for _, sweep := range purgeList {
+		b.returnChan <- sweep
+	}
+
+	b.log.Infof("spent, total sweeps: %v, purged sweeps: %v", spendCounter,
+		len(purgeList))
+
+	b.wg.Add(1)
+	go func() {
+		defer b.wg.Done()
+		b.monitorConfirmations(ctx)
+	}()
+
+	return b.persist(ctx)
+}
+
+// handleConf handles a confirmation notification. This is the final step of the
+// batch. Here we signal to the batcher that this batch was completed.
+func (b *Batch) handleConf(ctx context.Context) error {
+	b.log.Infof("confirmed")
+
+	b.State = Confirmed
+
+	b.completeChan <- b.ID
+
+	return b.markConfirmed(ctx)
+}
+
+// markConfirmed marks the batch as confirmed in the database.
+func (b *Batch) markConfirmed(ctx context.Context) error {
```

Comment:
Feels like this function is not necessary and could be inlined as it is only called from the function just above.